Refactoring topology name fixup of Intel mach

brentlu opened the pull request:
This pull request tries to use the code which fixup I2S topology file name to fixup HDA/SDW topology as well. Then redundant code could be removed.

brentlu committed:
ASoC: SOF: Intel: hda: refactoring topology name fixup for HDA mach

Move I2S mach's topology name fixup code to the end of machine driver
enumeration flow so HDA mach could also use same code to fixup its
topology file name as well. No functional change in this commit.

Signed-off-by: Brent Lu <brent.lu@intel.com>

plbossart commented on include/sound/soc-acpi-intel-match.h:
```diff
@@ -51,6 +51,9 @@ extern struct snd_soc_acpi_mach snd_soc_acpi_intel_ptl_sdw_machines[];
  * additional ACPI-enumerated codecs
  */
+#define SND_SOC_ACPI_INTEL_HDA_MACH_SKL	0
+#define SND_SOC_ACPI_INTEL_HDA_MACH_SOF	1
```
do we need to do this?

Most of the SKL/KBL platforms with the DSP are supposed to use the AVS driver now.

Rather than deal with differences with an obsolete driver, maybe we should remove support for HDaudio in SKL?

@crojewsk what do you think?

crojewsk-intel replied:
Sorry for the delay. I'll take a look today and respond either today or tomorrow. Thank you @plbossart for informing me about this.

crojewsk-intel replied:
ACK for ignoring skylake-driver here.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1310,12 +1296,46 @@ struct snd_soc_acpi_mach *hda_machine_select(struct snd_sof_dev *sdev)
+#endif
+
+	/* fixup tplg name for I2S and HDA machine driver */
+	if (!sdw_mach_found) {
```
and why not for SoundWire as well?

plbossart reviewed (commented):
Sorry, not following this at all, it's a lot of code changes and I don't really see what you are trying to fix.

brentlu commented:
There is no bug/issue here. I just try to use same code to fixup topology name of I2S/HDA machine drivers. And yes, we could use same code for SDW as well. Let me upload an new version after local testing. Thanks.

brentlu committed:
ASoC: SOF: Intel: hda: refactoring topology name fixup for SDW mach

Remove SDW mach's topology name fixup code and use the code in
hda_machine_select() to fixup its topology file name. No functional
change in this commit.

Compared with I2S/HDA mach, SDW mach always fixup topology file name
with dmic num without using DMIC quirk flag and pass topology name
with file extension to SOF driver. Therefore, we add extra code to
remove file extension if it exists.

Signed-off-by: Brent Lu <brent.lu@intel.com>

brentlu commented:
I've uploaded new patches which are focusing on the topology name fixup and minimizing each patch. Now all mach could use same code to handle DMIC name fixup (-dmicXch or -Xch). The flow to enumerate mach is not touched. Only the code related to topology name fixup.

plbossart commented:
Not sure what's going on but we've got a nasty TGL IPC timeout I don't recall having seen before:

https://sof-ci.01.org/linuxpr/PR5037/build3377/devicetest/index.html?model=TGLU_UP_HDA-ipc4&testcase=check-signal-stop-start-capture-10

brentlu commented:
> Not sure what's going on but we've got a nasty TGL IPC timeout I don't recall having seen before:
> 
> https://sof-ci.01.org/linuxpr/PR5037/build3377/devicetest/index.html?model=TGLU_UP_HDA-ipc4&testcase=check-signal-stop-start-capture-10

This PR only touch topology name string. Should not result in runtime IPC error. Maybe we could rerun the test?

plbossart commented:
SOFCI TEST

plbossart commented:
SOFCI TEST

brentlu commented:
Hi @plbossart @bardliao , is this PR looking good to you?

bardliao commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1071,33 +1068,30 @@ static void hda_generic_machine_select(struct snd_sof_dev *sdev,
-				return;
 
-			hda_mach->mach_params.dmic_num = dmic_num;
-			pdata->tplg_filename = tplg_filename;
```
Don't we need to assign pdata->tplg_filename?

brentlu replied:
We need to keep it NULL because caller (hda_machine_select()) will check this variable again for the value of its tplg_fixup local variable. So here we update the file name string in mach.

plbossart commented on sound/soc/intel/common/soc-acpi-intel-hda-match.c:
```diff
@@ -30,6 +30,11 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_hda_machines[] = {
+		/* SOF HDA machine driver */
+		.drv_name = "skl_hda_dsp_generic",
+		.sof_tplg_filename = "sof-hda-generic.tplg",
+	},
```
I am still not very hot on this, I am not sure if this is required, let's ping @crojewsk @crojewsk-intel again on whether SKL is still in use with HDaudio.

crojewsk-intel replied:
To my best knowledge, starting with v5.15 there is not a single system utilizing the skylake-driver regardless of the configuration we would be speaking of. I probably should have sorted out the dsp-driver selection on upstream earlier as we're in v6.10 already. Will prepare something either this or next week.

ACK for ignoring the skylake-driver here.

plbossart replied:
sounds good thanks @crojewsk-intel 

@brentlu we can simplify this assuming only SOF uses this machine driver.

brentlu replied:
I've uploaded new patch which removes HDA external codec support from Skylake driver.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -558,7 +558,7 @@ static int check_nhlt_ssp_mclk_mask(struct snd_sof_dev *sdev, int ssp_num)
 }
 
-#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA_AUDIO_CODEC) || IS_ENABLED(CONFIG_SND_SOC_SOF_INTEL_SOUNDWIRE)
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_INTEL_SOUNDWIRE)
```
nothing below is SoundWire specific, so it looks like a miss in terms of refactoring?

brentlu replied:
Two functions fixup_tplg_name() and dmic_detect_topology_fixup() are needed by hda_sdw_machine_select() so we have to keep the kernel config check here. They will be removed when refactoring hda_sdw_machine_select() function.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1348,7 +1366,7 @@ struct snd_soc_acpi_mach *hda_machine_select(struct snd_sof_dev *sdev)
 						       "%s%s%d%s",
 						       sof_pdata->tplg_filename,
-						       "-dmic",
+						       i2s_mach_found ? "-dmic" : "-",
```
that's almost the same code as line 580, so there's likely additional simplifications possible.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1199,45 +1125,10 @@ static struct snd_soc_acpi_mach *hda_sdw_machine_select(struct snd_sof_dev *sdev
-			"SoundWire machine driver %s topology %s\n",
-			mach->drv_name,
-			pdata->tplg_filename);
-
```
ah, I was right, here's the simplification.

to make the review easier, I would add a comment in the previous patch that this entire section will be removed in a follow-up optimization.

plbossart reviewed (commented):
I am afraid we missed the train for 6.11 while we were busy with other things. This will have to wait a bit @brentlu before being merged.

brentlu committed:
ASoC: Intel: Skylake: remove HDA machine driver support

Currently both Skylake and SOF platform driver enumerate same HDA
machine driver 'skl_hda_dsp_generic' for HDA external codec. This
commit removes HDA external codec support on Skylake platform so the
machine driver will be used by SOF only.

Signed-off-by: Brent Lu <brent.lu@intel.com>

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1333,15 +1241,20 @@ struct snd_soc_acpi_mach *hda_machine_select(struct snd_sof_dev *sdev)
 		 */
 		if (!sof_pdata->tplg_filename) {
-			sof_pdata->tplg_filename = mach->sof_tplg_filename;
+			/* remove file extension if exist */
```
if is exists

brentlu replied:
fixed

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1359,8 +1272,24 @@ struct snd_soc_acpi_mach *hda_machine_select(struct snd_sof_dev *sdev)
 		/* report to machine driver if any DMICs are found */
 		mach->mach_params.dmic_num = check_dmic_num(sdev);
 
+		/* SDW mach does not use DMIC quirk flag. */
```
I can't figure out what this comment means, and same for the commit message

SDW mach does not use the DMIC quirk flag and pass full topology name
with file extension to SOF driver. Therefore, we add extra code to
hda_generic_machine_select() function to cope with the difference.

if the goal was to unify, why do we still need differences with SoundWire handling for the dmic case? it's the same in the end, we need the -2ch or -4ch suffix appended, no?

brentlu replied:
Sorry my wording is misleading. In this PR I am trying to use same code to do topology name fixup for all I2S/HDA/SDW mach. There is no functional change in these commits. I'm updating the git commit/comment for more accurate description.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -511,6 +511,8 @@ static int check_dmic_num(struct snd_sof_dev *sdev)
 		dmic_num = intel_nhlt_get_dmic_geo(sdev->dev, nhlt);
 
+	dev_info(sdev->dev, "DMICs detected in NHLT tables: %d\n", dmic_num);
+
```
typo in commit message: extenion

brentlu replied:
fixed

plbossart reviewed (approved):
@bardliao can you please take a look?

marc-hb commented:
> > if the goal was to unify, why do we still need differences with SoundWire handling for the dmic case? it's the same in the end, we need the -2ch or -4ch suffix appended, no?

>  There is no functional change in these commits. 

Are you sure? The firmware just stopped loading in some configurations with 2 DMICs:
- #5124

brentlu commented:
> > > if the goal was to unify, why do we still need differences with SoundWire handling for the dmic case? it's the same in the end, we need the -2ch or -4ch suffix appended, no?
> 
> > There is no functional change in these commits.
> 
> Are you sure? The firmware just stopped loading in some configurations with 2 DMICs:
> 
> * [sof_sdw: init platform error / mc_probe warning #5124](https://github.com/thesofproject/linux/issues/5124)

Hmm, this PR changes to code to compose "topology file name" with reduced code. So the problem we will have is the new code generates a topology file name which does not exist. You will see kernel message prompts user to download the topology file from SOF github.
